File: 31. Next Permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int firstSmall = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i+1]) {
                firstSmall = i;
                break;
            }
        }
        
        if (firstSmall == -1) {
            reverse(nums,0,nums.length-1);
            return;
        }
        
        int firstLarge = -1;
        for(int j = nums.length - 1; j > firstSmall; j--) {
            if (nums[j] > nums[firstSmall]) {
                firstLarge = j;
                break;
            }
        }
        
        swap(nums,firstSmall, firstLarge);
        reverse(nums,firstSmall + 1, nums.length - 1);
        return;
    }
    
    
    public void reverse(int[] nums, int i, int j) {
        while(i < j) {
            swap(nums, i, j);
            i++;
            j--;
        }
    }
    
    
    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

// Time complexity: O(n)
// Space complexity: O(1)

File: python/31.Next Permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        N = len(nums)
        pivot = 0
        for i in range(N-1,0,-1):
            if nums[i-1] < nums[i]:
                pivot = i
                break 
        if pivot == 0:
            return nums.sort()
        
        swap = N - 1
        while nums[pivot-1] > nums[swap]:
            swap = swap - 1

        nums[pivot-1], nums[swap] = nums[swap], nums[pivot-1]
        nums[pivot:] = sorted(nums[pivot:])
        return nums 

# Time: O(N)
# Space: O(1)
